### PR TITLE
Feature/group fold button

### PR DESCRIPTION
**Pull request by ulferts**
Adds a button to fold/expand all groups displayed in the wp table. The button is only displayed when grouping (regardless of the grouping criteria).

Codewise this removes the extra listeners on the grouping states introduced in #8800. The timeline is currently always completely refreshed when a group is toggled. 

https://community.openproject.com/wp/35334

Also fixes https://community.openproject.com/wp/35359

### Todo

* [x] Check whether a drop down is easily possible
* [x] ~~Check for a potentially more fitting icon~~ checked but there isn't one. I opt for adding it later if it is deemed necessary.

### CODE REVIEW

**Comment by openprojectci** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/3bb6abe26dabf140f0427f138b1f0276d0674b1a/frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts#L68">frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts#L68</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by lindenthal** [comment]
Hi @ulferts,

I just had a quick look at the feature. Two remarks for now:

1. The position of the button is unclear. I would rather move it left to the view selector.
2. The texts and icons are swapped.
3. There is an icon outline.svg, which we could us. We could also try to remove the text label.

The status is not saved in the query, right?

Best
Niels

**Comment by ulferts** [comment]
@lindenthal I have made the requested changes. 

Removing the label looks weird in my opinion but if you think otherwise, I can make that change:

image

> The status is not saved in the query, right?

That is correct

**Comment by ulferts** [comment]
As discussed with @lindenthal, the button no longer has a text.

File: frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts [hunk 1]
```diff
@@ -60,12 +60,12 @@ import {WorkPackageNotificationService} from 'core-app/modules/work_packages/not
 import {combineLatest, merge, Observable} from 'rxjs';
 import {UntilDestroyedMixin} from 'core-app/helpers/angular/until-destroyed.mixin';
 import {WorkPackagesTableComponent} from 'core-components/wp-table/wp-table.component';
-import {GroupObject} from 'core-app/modules/hal/resources/wp-collection-resource';
 import {SchemaCacheService} from 'core-components/schemas/schema-cache.service';
 import {
   groupIdFromIdentifier,
   groupTypeFromIdentifier
 } from 'core-components/wp-fast-table/builders/modes/grouped/grouped-rows-helpers';
+import {WorkPackageViewCollapsedGroupsService} from "core-app/modules/work_packages/routing/wp-view-base/view-services/wp-view-collapsed-groups.service";
```
Comment (openprojectci) [inline frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts:69]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/ef92c6ebaffca32e102e33966ccc35b9fa7e258e/frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts#L69"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

**Comment by aleixsuau** [comment]
Hi @ulferts ,

I just pushed the following changes:

- Extracts all the grouping functionality to the WorkPackageViewCollapsedGroupsService which acts as a bridge between the collapse buttons (GroupRowHandler and WorkPackageGroupToggleDropdownMenuDirective) and the GroupFoldTransformer, which connects to the table itself.
- Rendering only the group rows when the grouping changes.
- Keep the state and remove wp icons from group rows when the grouping changes (ie: from project to status).
- Handle disable/enable grouping menu options.

The changes follow your approach of sticking to the actual way the table works. The main issue that I see is that every time we want to react to the table changes, we have to insert new code into it (GroupRowHandler, GroupFoldTransformer). I think it could be better to create a 'TableService' that would expose a clear API (e.g: this.tableService.groups.setAllGroupsCollapseStateTo(true)) than having to add services inside transformers inside the table, but this would require a bigger refactor not suited for this task.

Please take a look and let me know what you think.

Thanks!

File: frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts [hunk 2]
```diff
@@ -57,15 +57,14 @@ import {debugLog, timeOutput} from 'core-app/helpers/debug_output';
...
+import {combineLatest, Observable} from 'rxjs';
 import {UntilDestroyedMixin} from 'core-app/helpers/angular/until-destroyed.mixin';
 import {WorkPackagesTableComponent} from 'core-components/wp-table/wp-table.component';
-import {GroupObject} from 'core-app/modules/hal/resources/wp-collection-resource';
-import {SchemaCacheService} from 'core-components/schemas/schema-cache.service';
 import {
   groupIdFromIdentifier,
   groupTypeFromIdentifier
 } from 'core-components/wp-fast-table/builders/modes/grouped/grouped-rows-helpers';
+import {WorkPackageViewCollapsedGroupsService} from 'core-app/modules/work_packages/routing/wp-view-base/view-services/wp-view-collapsed-groups.service';
```
Comment (openprojectci) [inline frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts:68]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/23703d5b6f7116bc18de0bf8b07561a5ad8112be/frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts#L68"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

**Comment by ulferts** [comment]
Seems to work now, @aleixsuau . Thanks for stepping up.

**Pull request merged**